File: src/python/capture/camera.py
# capture/camera.py
import cv2
import time
import queue
from .base import CaptureThread
from utils.logger import get_logger

class CameraCapture(CaptureThread):
    def __init__(self, cam_id, output_q, config):
        super().__init__(output_q, config)
        self.cam_id = cam_id
        self.log = get_logger(f"Camera{cam_id}")
        self.cap = cv2.VideoCapture(cam_id)
        self._configure_camera()

    def _configure_camera(self):
        width = self.cfg.get("camera", {}).get("width", 640)
        height = self.cfg.get("camera", {}).get("height", 480)
        fps = self.cfg.get("camera", {}).get("fps", 30)
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
        self.cap.set(cv2.CAP_PROP_FPS, fps)
        if not self.cap.isOpened():
            raise RuntimeError(f"Camera {self.cam_id} could not be opened.")

    def run(self):
        while not self.stopped():
            ret, frame = self.cap.read()
            if not ret:
                self.log.warning("Frame grab failed.")
                time.sleep(0.01)
                continue
            ts = time.time()
            if self.q.full():
                try:
                    self.q.get_nowait()
                except queue.Empty:
                    pass
            self.q.put((ts, frame))

        self.cap.release()

File: src/cpp/src/capture/camera.cpp
// --- src/cpp/src/capture/camera.cpp ---------------------------------
#include "capture/camera.hpp"
#include <chrono>

using namespace zybo::capture;

CameraCapture::CameraCapture(int cam_id, int width, int height, int fps, std::queue<Frame>& out_q)
    : cam_id_{cam_id}, width_{width}, height_{height}, fps_{fps}, q_{out_q} {
    cap_.open(cam_id_);
    if (!cap_.isOpened()) {
        throw std::runtime_error("Camera open failed");
    }
    cap_.set(cv::CAP_PROP_FRAME_WIDTH, width_);
    cap_.set(cv::CAP_PROP_FRAME_HEIGHT, height_);
    cap_.set(cv::CAP_PROP_FPS, fps_);
}

CameraCapture::~CameraCapture() {
    stop();
}

void CameraCapture::start() {
    running_ = true;
    th_ = std::thread(&CameraCapture::run, this);
}

void CameraCapture::stop() {
    running_ = false;
    if (th_.joinable()) th_.join();
}

void CameraCapture::run() {
    const auto frame_period = std::chrono::milliseconds(static_cast<int>(1000 / fps_));
    while (running_) {
        cv::Mat frame;
        if (!cap_.read(frame)) {
            std::this_thread::sleep_for(frame_period);
            continue;
        }
        Frame f{static_cast<double>(cv::getTickCount()) / cv::getTickFrequency(), frame};
        if (q_.size() >= 2) q_.pop();
        q_.push(std::move(f));
        std::this_thread::sleep_for(frame_period);
    }
}